src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs

```diff
@@ -1,54 +1,129 @@
     {
-        var setAsDispatchedHolder = new SetAsDispatchedHolder { ContainerHolder = containerHolderResolver.ResolveAndSetIfAvailable(context) };
+        var setAsDispatchedHolder = new SetAsDispatchedHolder
+        {
+            ContainerHolder = containerHolderResolver.ResolveAndSetIfAvailable(context)
+        };
         context.Set(setAsDispatchedHolder);
 
-        if (!context.TryGet(out PartitionKey partitionKey))
+        var havePartitionKeyInContext = context.TryGet(out PartitionKey extractedPartitionKey);
+
+        PartitionKey finalPartitionKey = PartitionKey.Null;
+        bool shouldDeferToLogicalStage = false;
+
+        // Determine if we should defer to logical stage for partition key extraction
+        if (!havePartitionKeyInContext)
         {
-            // because of the transactional session we cannot assume the incoming message is always present
-            if (!context.TryGet<IncomingMessage>(out IncomingMessage incomingMessage) ||
-                !incomingMessage.Headers.ContainsKey(NServiceBus.Headers.ControlMessageHeader))
+            if (extractorConfig.HasCustomPartitionMessageExtractors)
+            {
+                // Custom partition key extractors need the message body
+                shouldDeferToLogicalStage = true;
+            }
+            else if (extractorConfig.HasCustomPartitionHeaderExtractors)
             {
-                // we return null here to enable outbox work at logical stage
-                return null;
+                // If we dont have a partition key here, but expect to via a custom header extractor, we need to throw
+                throw new Exception($"For the outbox to work a partition key must be provided either in the incoming physical or at latest in the logical message stage. Set one via '{nameof(CosmosPersistenceConfig.TransactionInformation)}'.");
             }
+            else if (!extractorConfig.HasAnyCustomPartitionExtractors)
+            {
+                // Use default synthetic partition key
+                finalPartitionKey = GetPartitionKey(extractedPartitionKey, messageId);
+            }
+        }
+        else
+        {
+            // We have a partition key from context (headers or previous extraction)
+            finalPartitionKey = extractedPartitionKey;
+        }
 
-            partitionKey = new PartitionKey(messageId);
-            context.Set(partitionKey);
+        // Check if we need to defer for container extraction
+        if (!setAsDispatchedHolder.ContainerIsSet() && extractorConfig.HasCustomContainerMessageExtractors)
+        {
+            // When deferring for container extraction, ensure partition key is set in context
+            // Use the default synthetic key if no custom partition extractors are configured
+            if (!havePartitionKeyInContext && !extractorConfig.HasAnyCustomPartitionExtractors)
+            {
+                finalPartitionKey = GetPartitionKey(extractedPartitionKey, messageId);
```

Comment:
I don't think this is needed as the finalPartitionKey is already set above to what it should be?